internal/cody/subscription.go

```diff
@@ -0,0 +1,147 @@
+package cody
+
+import (
+	"context"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/featureflag"
+	"github.com/sourcegraph/sourcegraph/internal/ssc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+const USE_SSC_FEATURE_FLAG = "use-ssc-for-cody-subscription"
+
+type UserSubscriptionPlan string
+
+const (
+	UserSubscriptionPlanFree UserSubscriptionPlan = "free"
+	UserSubscriptionPlanPro  UserSubscriptionPlan = "pro"
+)
+
+type UserSubscription struct {
+	// Status is the current status of the subscription. "pending" means the user has no Cody Pro subscription.
+	// "pending" subscription will be removed post Feb 15, 2024. It is required to support users who have opted for
+	// Cody Pro Trial on dotcom.
+	Status ssc.SubscriptionStatus
+	// Plan is the plan the user is subscribed to. "free" or "pro".
+	Plan UserSubscriptionPlan
+	// ApplyProRateLimits is true if the user has a valid subscription status.
+	ApplyProRateLimits bool
+	// CurrentPeriodStartAt is the start date of the current billing cycle.
+	CurrentPeriodStartAt time.Time
+	// CurrentPeriodEndAt is the end date of the current billing cycle.
+	CurrentPeriodEndAt time.Time
+}
+
+// consolidateSubscriptionDetails consolidates the subscription details from dotcom and SSC.
+// Post Feb 15, 2024, this function will be removed and the logic will be simplified to us the SSC info only.
```

Comment:
Super-minor suggestion: Your comment was good enough, but I would (personally, by preference) word it slightly differently.

```suggestion
// consolidateSubscriptionDetails merges the subscription data available on dotcom and SCC.
// This is needed while transitioning to use SCC as the source of truth for all subscription data. (Which should happen ~Q1/2024.)
```

Also, could you please file a ticket to track actually doing this work? That way we can replace the bottom line with the following. (So we can actually track and follow-through with our plan here.)

```go
// TODO[sourcegraph#12345]: Update dotcom to use SSC as the source of truth for all subscription data.
```

Reply:
https://github.com/sourcegraph/sourcegraph/issues/59785